[13.0] Fix groups of transfers on sales

guewen opened the pull request:
When grouping transfers of different sales orders in the same
procurement.group, we have an issue with only the first sales order
showing the "Transfers" button on the SO view, because
"sale.picking_ids" returns an empty recordset on the other sales orders.

This is because the link between sales and pickings is done through the
"group_id.sale_id" field of the stock moves. When we group transfers of
different sales, we end up with several group_id in the moves of the
same picking, which is a bit unexpected, and when new transfers are
created (dynamic routing, backorders, ...), they'll only get the first
group_id and be attached only to the first sale order.

The solution implemented here is the following:

* Add a M2m "sale_ids" field on the procurement group
* When several moves with different groups are merged into the same
  group, clone the group, update its "sale_ids" field to contain all the
  sales (sale_id will keep the first one though) and set in on the moves
* Update the computed fields for the relation between the sales and
  pickings to use this M2M field
* Keep the original group in the stock move as "original_group_id": used
  when a sales order is cancelled

Commit:
Fix groups of transfers on sales

When grouping transfers of different sales orders in the same
procurement.group, we have an issue with only the first sales order
showing the "Transfers" button on the SO view, because
"sale.picking_ids" returns an empty recordset on the other sales orders.

This is because the link between sales and pickings is done through the
"group_id.sale_id" field on the stock moves. When we group transfers of
different sales, we end up with several group_id in the moves of the
same picking, which is a bit unexpected, and when new transfers are
created (dynamic routing, backorders, ...), they'll only get the first
group_id and be attached only to the first sale order.

The solution implemented here is the following:

* Add a M2m "sale_ids" field on the procurement group
* When several moves with different groups are merged into the same
  group, clone the group, update its "sale_ids" field to contain all the
  sales (sale_id will keep the first one though) and set in on the moves
* Update the computed fields for the relation between the sales and
  pickings to use this M2M field
* Keep the original group in the stock move as "original_group_id": used
  when a sales order is cancelled

Commit:
Simplify line group comparison

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -1,34 +1,76 @@
+    # store the first group the move was in when created, used for cancellation
+    # from a sales order
+    original_group_id = fields.Many2one(
+        comodel_name="procurement.group", string="Procurement Group",
```
```suggestion
        comodel_name="procurement.group", string="Procurement Group Original",
```

jbaudoux replied:
To prevent having 2 fields on stock.move with the same name

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -1,34 +1,76 @@
+            if not picking.picking_type_id.group_pickings:
+                return
+            base_group = picking.group_id
+            base_sales = base_group.sale_ids
```
This can contain sales that are already done. For example procurement group is linked to SO1 & SO2. SO1 is delivered. Back order contains only SO2 but procurement group is still linked to SO1 & SO2. When SO3 occurs, you have in the backorder SO2 & SO3 but the base group contains SO1 & SO2. The new group should only contain SO2 & SO3 without SO1. Otherwise, on the link from SO1 you see too much pickings (and all the one coming in the future) and the proc group link and name will grow too much.

Commit:
Prevent moves being always merged in the same groups

If we have a unique transfer for 2 sales orders, we have one joint
procurement group for them. If we delivery the first sale order, we
expect to no longer have the first SO in the group of the backorder.

This will prevent to have a never-ending group with all the orders from
the beginning if we have many backorders.

To so do, we consider that a "printed" or done transfer will no longer
change and therefore, we consider its group frozen.

Commit:
Rename/move around methods for clarity

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -26,34 +28,40 @@ def _assign_picking_post_process(self, new=False):
 
-    def _prepare_merge_group_values(self, sales):
+    def _prepare_merge_group_values(self, move_groups):
+        sales = move_groups.sale_id
```
mapped?

guewen replied:
no longer necessary on relations in 13.0 :)

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -26,34 +28,40 @@ def _assign_picking_post_process(self, new=False):
-            line_sales = moves.group_id.sale_id
-            all_sales = base_sales | line_sales
+            moves = picking.move_lines
+            moves_groups = moves.original_group_id
```
mapped?

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -26,34 +28,40 @@ def _assign_picking_post_process(self, new=False):
-            all_sales = base_sales | line_sales
+            moves = picking.move_lines
+            moves_groups = moves.original_group_id
+            moves_sales = moves_groups.sale_id
```
mapped?

yvaucher commented:
/ocabot merge nobump
